### PR TITLE
#668: added a type constraint to provoke an error message

### LINKED ISSUE #668: Missing error apparently caused by erroneous type inference
class X {
  function method F(): int { 2 }
}

class K { }

method P0(x: X, k: K) {
  var o;  // type inferred to be object
  o := x;
  o := k;
  var u := o.F();  // F does not exist in object
}

With Rustan's PR https://github.com/dafny-lang/dafny/pull/667, this program checks without errors! That’s bad.

### PR DESCRIPTION
**Pull request by davidcok**
Fixes #668

### CODE REVIEW

File: Source/Dafny/Resolver.cs [hunk 1]
```diff
@@ -14427,6 +14427,7 @@ public class ResolveOpts
           Expression receiver;
           if (!member.IsStatic) {
             receiver = expr.Lhs;
+            AddAssignableConstraint(expr.tok, tentativeReceiverType, receiver.Type, "receiver type does not have a member named " + name);
```
Comment (RustanLeino) [inline Source/Dafny/Resolver.cs:14430]: Add name of receiver type to error message. You can you do simply as follows:

            AddAssignableConstraint(expr.tok, tentativeReceiverType, receiver.Type, "receiver type ({1}) does not have a member named " + name);

File: Test/git-issues/git-issue-668.dfy [hunk 2]
```diff
@@ -0,0 +1,32 @@
+// RUN: %dafny /compile:0 "%s" > "%t"
+// RUN: %diff "%s.expect" "%t"
+
+class X {
+  function method F(): int { 2 }
+  var J: int;
```
Comment (RustanLeino) [inline Test/git-issues/git-issue-668.dfy:6]: remove semi-colon

Comment (davidcok) [reply Test/git-issues/git-issue-668.dfy]: Done

File: Source/Dafny/Resolver.cs (hunk 1 above)
Comment (davidcok) [reply Source/Dafny/Resolver.cs]: Done

**Pull request merged**